File: apps/admin/src/modules/devices/chat/chat-sessions.ts
import { randomUUID } from 'crypto';
import type WebSocket from 'ws';
import {
  CHAT_SESSION_LIMIT, DIRECT_CONNECT_TIMEOUT, identifier, publicKey, send, signal,
  type ChatIdentity, type ChatSession,
} from './protocol';

/** Rendezvous and opaque fallback relay. Conversation data is never stored here. */
export class ChatSessions {
  private readonly desktops = new Map<string, WebSocket>();
  private readonly sessions = new Map<string, ChatSession>();

  join(client: WebSocket, identity: ChatIdentity, message: Record<string, unknown>, iceServers: object[]) {
    const key = `${identity.ownerId}:${identity.deviceId}`;
    if (identity.role === 'desktop') {
      const previous = this.desktops.get(key);
      if (previous && previous !== client) {
        this.disconnect(previous);
        previous.close(4000, 'Replaced by a newer connection');
      }
      this.desktops.set(key, client);
      send(client, { type: 'registered' });
      return;
    }
    const desktop = this.desktops.get(key);
    if (!desktop || desktop.readyState !== 1) {
      client.close(4004, 'PC chat is offline');
      return;
    }
    if ([...this.sessions.values()].filter((entry) => entry.desktop === desktop).length >= CHAT_SESSION_LIMIT) {
      client.close(4008, 'Too many chat connections');
      return;
    }
    const peerKey = publicKey(message.publicKey);
    const id = randomUUID();
    this.sessions.set(id, { id, desktop, mobile: client, startedAt: Date.now(), relay: false });
    send(desktop, { type: 'peer-open', sessionId: id, publicKey: peerKey, iceServers });
    send(client, { type: 'paired', sessionId: id, iceServers, directTimeoutMs: DIRECT_CONNECT_TIMEOUT });
  }

  route(client: WebSocket, message: Record<string, unknown>) {
    const sessionId = identifier(message.sessionId);
    const session = this.sessions.get(sessionId);
    if (!session || (client !== session.desktop && client !== session.mobile)) throw new Error('Unknown session');
    const target = client === session.desktop ? session.mobile : session.desktop;
    if (message.type === 'signal') {
      send(target, { type: 'signal', sessionId, payload: signal(message.payload) });
      return;
    }
    if (message.type === 'relay-request') {
      // Initial fallback is allowed only after the direct attempt. Later transport failures can fall back immediately.
      if (!session.relay && message.reason !== 'disconnected'
        && Date.now() - session.startedAt < DIRECT_CONNECT_TIMEOUT) throw new Error('Direct attempt still pending');
      session.relay = true;
      send(session.desktop, { type: 'relay-ready', sessionId });
      send(session.mobile, { type: 'relay-ready', sessionId });
      return;
    }
    if (message.type === 'relay' && session.relay && typeof message.payload === 'string'
      && /^[a-f0-9]+$/.test(message.payload) && message.payload.length <= 40_000) {
      send(target, { type: 'relay', sessionId, payload: message.payload });
      return;
    }
    if (message.type === 'peer-close') {
      this.sessions.delete(sessionId);
      send(target, { type: 'peer-close', sessionId });
      return;
    }
    throw new Error('Invalid chat frame');
  }

  disconnect(client: WebSocket) {
    for (const [key, desktop] of this.desktops) if (desktop === client) this.desktops.delete(key);
    for (const [id, session] of this.sessions) {
      if (session.desktop !== client && session.mobile !== client) continue;
      this.sessions.delete(id);
      const target = session.desktop === client ? session.mobile : session.desktop;
      send(target, { type: 'peer-close', sessionId: id });
    }
  }
}

File: apps/admin-go/internal/devices/chat_sessions.go
package devices

import (
	"errors"
	"sync"
	"time"

	"github.com/codex-switch/admin-go/internal/platform"
	"github.com/google/uuid"
)

type desktopInfo struct {
	expires time.Time
	version float64
}
type chatSessions struct {
	mu       sync.Mutex
	desktops map[string]*peer
	info     map[*peer]desktopInfo
	sessions map[string]*chatSession
	hot      *hotSessions
	onRelay  func(int)
	deliver  func(relayDelivery, platform.JSON)
}

func newChatSessions(relay func(int)) *chatSessions {
	return &chatSessions{desktops: map[string]*peer{}, info: map[*peer]desktopInfo{},
		sessions: map[string]*chatSession{}, hot: newHotSessions(relay), onRelay: relay}
}

func (s *chatSessions) setLimit(limit float64) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.hot.limit = limit
}

func (s *chatSessions) join(client *peer, identity chatIdentity, message platform.JSON, ice []platform.JSON) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.hot.prune()
	key := identity.owner + ":" + identity.device
	version := desktopTransportVersion(message["transportVersion"])
	if identity.role == "desktop" {
		var descriptors interface{}
		if message["transportVersion"] == float64(2) {
			var present bool
			descriptors, present = message["sessions"]
			if present && descriptors == nil {
				return errors.New("invalid sessions")
			}
		}
		if err := s.hot.register(client, identity, descriptors); err != nil {
			return err
		}
		s.info[client] = desktopInfo{identity.expires, version}
		if previous := s.desktops[key]; previous != nil && previous != client {
			s.disconnectLocked(previous, false)
			previous.close(4000, "Replaced by a newer connection")
		}
		s.desktops[key] = client
		client.send(platform.JSON{"type": "registered"}, nil)
		return nil
	}
	return s.joinMobile(client, identity, message, ice)
}

func (s *chatSessions) joinMobile(
	client *peer,
	identity chatIdentity,
	message platform.JSON,
	ice []platform.JSON,
) error {
	desktop := s.desktops[identity.owner+":"+identity.device]
	if desktop == nil || desktop.closed.Load() {
		client.close(4004, "PC chat is offline")
		return nil
	}
	count := s.hot.count(identity)
	for _, session := range s.sessions {
		if session.desktop == desktop {
			count++
		}
	}
	_, resuming := message["resume"]
	if !resuming && float64(count) >= s.hot.limit {
		client.close(4008, "Too many chat connections")
		return nil
	}
	info := s.info[desktop]
	if message["transportVersion"] == float64(2) && info.version == 2 {
		return s.hot.join(hotJoin{client, identity, message, chatEndpoint{desktop, info.expires}, ice})
	}
	if resuming {
		client.close(4004, "Session unavailable")
		return nil
	}
	key, err := publicKey(message["publicKey"])
	if err != nil {
		return err
	}
	id := uuid.NewString()
	s.sessions[id] = &chatSession{id: id, owner: identity.owner, desktop: desktop, mobile: client, started: time.Now()}
	desktop.send(withChatClientInfo(platform.JSON{"type": "peer-open", "sessionId": id,
		"publicKey": key, "iceServers": ice}, message["clientInfo"]), nil)
	client.send(
		platform.JSON{
			"type":            "paired",
			"sessionId":       id,
			"iceServers":      ice,
			"directTimeoutMs": directConnectTimeout.Milliseconds(),
		},
		nil,
	)
	return nil
}

func (s *chatSessions) route(client *peer, message platform.JSON) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	handled, err := s.hot.route(client, message)
	if handled || err != nil {
		return err
	}
	id, err := identifier(message["sessionId"])
	if err != nil {
		return err
	}
	session := s.sessions[id]
	if session == nil || (session.desktop != client && session.mobile != client) {
		return errors.New("unknown session")
	}
	target := session.desktop
	if client == session.desktop {
		target = session.mobile
	}
	switch message["type"] {
	case "signal":
		payload, err := chatSignal(message["payload"])
		if err != nil {
			return err
		}
		target.send(platform.JSON{"type": "signal", "sessionId": id, "payload": payload}, nil)
	case "relay-request":
		return s.enableRelay(session, message)
	case "relay":
		payload, valid := relayPayload(message["payload"])
		if !session.relay || !valid {
			return errors.New("invalid relay")
		}
		frame := platform.JSON{"type": "relay", "sessionId": id, "payload": payload}
		if s.deliver != nil {
			s.deliver(relayDelivery{session.owner, id, client, target, &session.traffic, client == session.desktop}, frame)
		} else {
			target.send(frame, s.onRelay)
		}
	case "peer-close":
		delete(s.sessions, id)
		target.send(platform.JSON{"type": "peer-close", "sessionId": id}, nil)
	default:
		return errors.New("invalid chat frame")
	}
	return nil
}

func (s *chatSessions) enableRelay(session *chatSession, message platform.JSON) error {
	if !session.relay && message["reason"] != "disconnected" && time.Since(session.started) < directConnectTimeout {
		return errors.New("direct attempt still pending")
	}
	session.relay = true
	frame := platform.JSON{"type": "relay-ready", "sessionId": session.id}
	session.desktop.send(frame, nil)
	session.mobile.send(frame, nil)
	return nil
}

func (s *chatSessions) disconnect(client *peer, revoke bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.disconnectLocked(client, revoke)
}

func (s *chatSessions) disconnectLocked(client *peer, revoke bool) {
	s.hot.disconnect(client, revoke)
	delete(s.info, client)
	for key, desktop := range s.desktops {
		if desktop == client {
			delete(s.desktops, key)
		}
	}
	for id, session := range s.sessions {
		if session.desktop != client && session.mobile != client {
			continue
		}
		delete(s.sessions, id)
		target := session.desktop
		if target == client {
			target = session.mobile
		}
		target.send(platform.JSON{"type": "peer-close", "sessionId": id}, nil)
	}
}

func (s *chatSessions) prune() { s.mu.Lock(); defer s.mu.Unlock(); s.hot.prune() }
